mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulEditProfileNameDialog.kt

```diff
@@ -0,0 +1,75 @@
+package ch.epfl.sdp.mobile.state
+
+import androidx.compose.foundation.MutatePriority
+import androidx.compose.foundation.MutatorMutex
+import androidx.compose.runtime.*
+import ch.epfl.sdp.mobile.application.authentication.AuthenticatedUser
+import ch.epfl.sdp.mobile.ui.setting.EditProfileNameDialog
+import ch.epfl.sdp.mobile.ui.setting.EditProfileNameDialogState
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+/**
+ * Implementation of the [EditProfileNameDialogState] interface
+ *
+ * @param user the current [AuthenticatedUser].
+ * @param scope the coroutine scope.
+ * @param onSave the callback called after we saved the changes.
+ * @param onCancel the callback called when we click the cancel button.
+ */
+class EditProfileNameDialogStateImpl(
+    private val user: AuthenticatedUser,
+    private val scope: CoroutineScope,
+    onSaveAction: State<() -> Unit>,
+    onCancelAction: State<() -> Unit>
+) : EditProfileNameDialogState {
+  private val onSaveAction by onSaveAction
+  private val onCancelAction by onCancelAction
+
+  override var name by mutableStateOf(user.name)
+
+  /**
+   * A [MutatorMutex] which enforces mutual exclusion of update profile name requests. Performing a
+   * new request (by clicking the button) will cancel the currently pending request.
+   */
+  private val mutex = MutatorMutex()
+
+  override fun onSaveClick() {
+    scope.launch {
+      mutex.mutate(MutatePriority.UserInput) {
+        user.update { name(name) }
```

Comment:
[Question] Does name(name) change the user's name? If so maybe one of both could have a different name.

Reply:
Yes, `name(name)` sends a request on firebase to update the username.

Good input, it is not quite intuitive what this line of code does. I will rename the `user.name->name` to  `user.name->userName`.